libs/ballot-interpreter/src/hmpb-rust/timing_marks.rs

```diff
@@ -363,34 +393,148 @@ pub fn find_timing_mark_grid(
+    if let Some(above_rect_within_image) = rect_within_image
+        .offset(0, -rect_height)
+        .intersect(&image_rect)
+    {
+        let above_rect_sub_image = GenericImageView::view(
+            image,
+            above_rect_within_image.left() as u32,
+            above_rect_within_image.top() as u32,
+            above_rect_within_image.width(),
+            above_rect_within_image.height(),
+        );
+        white_pixel_count += above_rect_sub_image
+            .pixels()
+            .filter(|(_, _, luma)| luma.0[0] > threshold)
+            .count();
+        possible_white_pixel_count +=
+            above_rect_within_image.width() * above_rect_within_image.height();
+    }
+
+    // shift the rect area down by the height of the timing mark, then look for
+    // white pixels in the rectangle
+    if let Some(below_rect_within_image) = rect_within_image
+        .offset(0, rect_height)
+        .intersect(&image_rect)
+    {
+        let below_rect_sub_image = GenericImageView::view(
+            image,
+            below_rect_within_image.left() as u32,
+            below_rect_within_image.top() as u32,
+            below_rect_within_image.width(),
+            below_rect_within_image.height(),
+        );
+        white_pixel_count += below_rect_sub_image
+            .pixels()
+            .filter(|(_, _, luma)| luma.0[0] > threshold)
+            .count();
+        possible_white_pixel_count +=
+            below_rect_within_image.width() * below_rect_within_image.height();
+    }
+
+    let white_pixel_score = if possible_white_pixel_count == 0 {
+        0.0
+    } else {
+        white_pixel_count as f32 / possible_white_pixel_count as f32
+    };
+
+    Some(black_pixel_score * white_pixel_score)
+}
+
 /// Filters the bottom timing marks by examining the image to determine if the
 /// timing marks are actually present.
 pub fn find_actual_bottom_marks(
     complete_timing_marks: &Complete,
-    image: &GrayImage,
-    threshold: u8,
+    ballot_image: &BallotImage,
 ) -> Vec<Option<Rect>> {
+    const MIN_REQUIRED_MARK_SCORE: f32 = 0.5;
     complete_timing_marks
         .bottom_rects
         .par_iter()
         .map(|rect| {
-            let rect_within_image =
-                rect.intersect(&Rect::new(0, 0, image.width(), image.height()))?;
-
-            let rect_sub_image = GenericImageView::view(
-                image,
-                rect_within_image.left() as u32,
-                rect_within_image.top() as u32,
-                rect_within_image.width(),
-                rect_within_image.height(),
-            );
-            let rect_area = rect.width() * rect.height();
-            if rect_sub_image
-                .pixels()
-                .filter(|(_, _, luma)| luma.0[0] <= threshold)
-                .count()
-                > (rect_area / 2) as usize
-            {
+            let score = score_timing_mark_at_rect(ballot_image, rect)?;
+            if score > MIN_REQUIRED_MARK_SCORE {
```

Comment:
Pedantic nit that probably doesn't matter in practice. If this is the "min required" score I would expect `>=` here